### PR TITLE
Preview of celo4 branch (do not merge)

### PR DESCRIPTION
**Pull request by karlb**
I have done a few changes to the code during the rebase to make it work. Please have a look at the comments I created and the following commits:
* https://github.com/celo-org/optimism/pull/100/commits/ba6b969af004a8a982aff0e98828cf345f473dd1
* https://github.com/celo-org/optimism/pull/100/commits/3bedc9674a114950d199b95f2a171e65803fd4b6
* https://github.com/celo-org/optimism/pull/100/commits/61982e7c7b44d25a29c29b21ddc7c0f89946b611

The rest should just be merging, squashing and minor cleanups.

### CODE REVIEW

File: op-chain-ops/genesis/helpers.go [hunk 1]
```diff
@@ -74,7 +74,14 @@ func AddressToCodeNamespace(addr common.Address) (common.Address, error) {
 }
 
 func IsL2DevPredeploy(addr common.Address) bool {
-	return bytes.Equal(addr[0:2], []byte{0x42, 0x00})
+	// Celo: We allow all predeploys here so that we can deploy contracts with
+	// proxies at arbitrary addresses. Since only the last two bytes are kept
+	// when mapping the address into `codeNamespace`, there is a 2^16 risk of
+	// collision. This is why Optimism only allowed that for addresses starting
+	// with 0x4200, because those are all chosen to differ in the last two
+	// bytes.
```
Comment (karlb) [inline op-chain-ops/genesis/helpers.go:82]: This was already the case before the rebase, I just added some documentation to explain this. See also `AddressToCodeNamespace` right before this.

File: packages/contracts-bedrock/src/celo/common/interfaces/IExchange.sol
Comment (karlb) [inline packages/contracts-bedrock/src/celo/common/interfaces/IExchange.sol:1]: This file was unused and has been removed as part of the rebase.

File: packages/contracts-bedrock/src/celo/FeeHandler.sol [hunk 2]
```diff
@@ -21,7 +21,7 @@ import "./common/interfaces/ICeloToken.sol";
 import "./stability/interfaces/ISortedOracles.sol";
 
 // Using the minimal required signatures in the interfaces so more contracts could be compatible
-import "./common/libraries/ReentrancyGuard.sol";
+import { ReentrancyGuard } from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
```
Comment (karlb) [inline packages/contracts-bedrock/src/celo/FeeHandler.sol:24]: We already used the ReentrancyGuard directly from OpenZepplin in another place and the tooling had problems with to copies of it, so I changed this to import directly from OpenZepplin, too.

Comment (palango) [reply packages/contracts-bedrock/src/celo/FeeHandler.sol:24]: Sounds reasonable to me, maybe something to check for @pahor167 . Maybe we can get completely rid of our internal version.

**Review by palango — APPROVED** [review]
Looks good to me. The comments and mentioned commits seem reasonable.

**Pull request closed without merging**